src/rvsdg/egglog_optimizer/conditional_invariant_code_motion.egg

```diff
@@ -0,0 +1,50 @@
+; If (= x (prefix-containing-operand (Gamma ...) e)), then the first x outputs
+; of Gamma contain e.
+(function prefix-containing-Operand (Body Operand) i64 :merge (max old new))
+(rule
+    ((= gamma (Gamma pred inputs outputs))
+     (Body-contains-Operand gamma 0 e))
+    ((set (prefix-containing-Operand gamma e) 0))
+    :ruleset fast-analyses)
+(rule
+    ((= gamma (Gamma pred inputs outputs))
+     (= i (prefix-containing-Operand gamma e))
+     (Body-contains-Operand gamma i e))
+    ((set (prefix-containing-Operand gamma e) (+ i 1)))
+    :ruleset fast-analyses)
+
+; Limit blowup by only repeatedly hoisting so many times and by limiting the
+; number of branches we hoist from or max number of inputs we can add.
+(let max-hoists 1)
+(function hoist-count (Body) i64 :merge (max old new))
+(rule ((= gamma (Gamma pred inputs outputs)))
+      ((set (hoist-count gamma) 0))
+      :ruleset fast-analyses)
+(let max-branches 2)
+(let max-inputs 4)
+
+; Hoist code out of a gamma, adding it as an argument instead.
+(rule 
+    ((= gamma (Gamma pred (VO inputs) outputs))
+     ; All branches contain hoist-me
+     (= (prefix-containing-Operand gamma hoist-me)
+        (VecVecOperand-length outputs))
+     (!= hoist-me (Arg any))
+     ; Can't add another version of an impure node since we might extract both.
+     ; Note: if we lift this requirement via an extraction-like substitution
+     ; procedure, then remember to add (Operand-is-pure hoist-me)!
+     (Body-is-pure gamma)
```

Comment:
Why does `gamma` (and `hoist-me`) has to be true? Even if `hoist-me` is not true, couldn't you still push them out?

Reply:
Do you mean pure?

I think hoist-me has to be pure, because can-subst-beneath is allowed to replace some but not all, but I think we can drop the requirement that gamma is pure.

Reply:
Makes sense. Thanks!